packit/config.py

```diff
@@ -168,21 +160,64 @@ def get_from_dict(cls, raw_dict: dict, validate=True) -> "Config":
             "command_handler_k8s_namespace", "myproject"
         )
 
+        config.services = Config.load_authentication(raw_dict)
         return config
 
-    @property
-    def github_token(self) -> str:
-        token = os.getenv("GITHUB_TOKEN", "")
-        if token:
-            return token
-        return self._github_token
+    @staticmethod
+    def load_authentication(raw_dict):
+        services = set()
+        if "authentication" in raw_dict:
+            services = get_instances_from_dict(instances=raw_dict["authentication"])
+        else:
+            logger.warning(
+                "Please,"
+                "use 'authentication' key in the user configuration "
+                "to set tokens for GitHub and Pagure. "
+                "New method supports more services and direct keys will be remove in the future.\n"
+                "Example:\n"
+                "authentication:\n"
+                "    github.com:\n"
+                "        token: GITHUB_TOKEN\n"
+                "    pagure:\n"
+                "        token: PAGURE_TOKEN\n"
+                '        instance_url: "https://src.fedoraproject.org"\n'
+            )
+            github_app_id = raw_dict.get("github_app_id")
+            github_app_cert_path = raw_dict.get("github_app_cert_path")
+            github_token = raw_dict.get("github_token")
+            services.add(
+                GithubService(
+                    token=github_token,
+                    github_app_id=github_app_id,
+                    github_app_private_key_path=github_app_cert_path,
+                )
+            )
+            pagure_user_token = raw_dict.get("pagure_user_token")
+            if raw_dict.get("pagure_fork_token"):
+                warnings.warn(
+                    "packit no longer accepts 'pagure_fork_token'"
+                    " value (https://github.com/packit-service/packit/issues/495)"
+                )
+            services.add(
+                PagureService(
+                    token=pagure_user_token,
+                    instance_url="https://src.fedoraproject.org",
+                )
+            )
 
-    @property
-    def pagure_user_token(self) -> str:
-        token = os.getenv("PAGURE_USER_TOKEN", "")
-        if token:
-            return token
-        return self._pagure_user_token
+        return services
+
+    def _get_project(self, url: str) -> GitProject:
+        try:
+            project = get_project(url=url, custom_instances=self.services)
+        except OgrException as ex:
+            msg = f"Authentication for url '{url}' is missing in the config."
+            logger.warning(msg)
+            raise PackitConfigException(msg, ex)
+        return project
+
+    def get_project(self, url: str) -> GitProject:
+        return Proxy(partial(self._get_project, url))
```

Comment:
The evaluation is postponed to the actual usage.

- Not all command (and tests) needs the project instance.
- Using `None` instead of the exception would cause misleading errors.